tooling/SecretManager.VisualStudio/ProjectLocalSecretsManager.cs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.Extensions.SecretManager.VisualStudio
+{
+    using Task = System.Threading.Tasks.Task;
+
+    public class ProjectLocalSecretsManager : IVsProjectSecrets, SVsProjectLocalSecrets
+    {
+        private const string UserSecretsPropertyName = "UserSecretsId";
+
+        private readonly IProjectProperties _projectProperties;
+
+        public ProjectLocalSecretsManager(IProjectProperties properties)
+        {
+            _projectProperties = properties;
+        }
+
+        public string SanitizeName(string name) => name;
+
+        public IReadOnlyCollection<char> GetInvalidCharactersFrom(string name) => Array.Empty<char>();
+
+        public Task AddSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+            => SetSecretAsync(name, value, cancellationToken);
+
+        public async Task SetSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValue(name);
+
+            var store = await GetOrCreateStoreAsync();
+
+            store.Set(name, value);
+
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                store.Save();
+            }
+        }
+
+        public async Task<string> GetSecretAsync(string name, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValue(name);
+            var store = await GetOrCreateStoreAsync();
+            return store[name];
+        }
+
+        public async Task<IReadOnlyCollection<string>> GetSecretNamesAsync(CancellationToken cancellationToken = default)
+        {
+            var store = await GetOrCreateStoreAsync();
+            return store.ReadOnlyKeys;
+        }
+
+        public async Task<IReadOnlyDictionary<string, string>> GetSecretsAsync(CancellationToken cancellationToken = default)
+        {
+            return await GetOrCreateStoreAsync();
+        }
+
+        public async Task<bool> RemoveSecretAsync(string name, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValue(name);
+
+            var store = await GetOrCreateStoreAsync();
+
+            var result = store.Remove(name);
+
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                store.Save();
```

Comment:
This will overwrite changes other (logical) threads may also be doing. For example, this:

```csharp
await (Task.WhenAll(store.AddAsync(1,2), store.AddAsync(3,4)));
```

Would very likely end up writing a file with only one of the secrets in it instead of both. We'd prefer the API to be more robust than that, and the interface documentation indicates the implementation is expected to be thread-safe.

Reply:
This will be problematic. I'll have to think about this more so we can avoid race conditions on reading from/writing to disk